src/branch_and_bound.rs

```diff
@@ -1,222 +1,785 @@
+//
+// * Addendum on Waste Calculation Optimization *
+// Waste, like accumulated value, is a bound used to track when a search path is no longer
+// advantageous.  The waste total is accumulated and stored in a variable called current_waste.
+// Besides the difference between amount and target, current_waste stores the difference between
+// utxo fee and utxo_long_term_fee.
+//
+// If the iteration adds a new node to the inclusion branch, besides incrementing the accumulated
+// value for the node, the waste is also added to the current_waste.  Note that unlike value,
+// waste can actually be negative.  This happens if there is a low fee environment such that
+// fee is less than long_term_fee.  Therefore, the only case where a solution becomes more
+// wasteful, and we may bound our search because a better waste score is no longer possible
+//
+//  a) We have already found a solution and the next solution is greater then the previous.
+//  b) It's a high fee environment such that adding more utxos will increase current_waste.
+//
+// If either a or b is true, we consider the search path no longer viable to continue down.
+pub fn select_coins_bnb(
+    target: Amount,
+    cost_of_change: Amount,
+    fee_rate: FeeRate,
+    long_term_fee_rate: FeeRate,
+    weighted_utxos: &mut [WeightedUtxo]
+) -> Option<std::vec::IntoIter<&WeightedUtxo>> {
+    // Total_Tries in Core:
+    // https://github.com/bitcoin/bitcoin/blob/1d9da8da309d1dbf9aef15eb8dc43b4a2dc3d309/src/wallet/coinselection.cpp#L74
+    const ITERATION_LIMIT: i32 = 100_000;
+
+    let mut iteration = 0;
+    let mut index = 0;
+    let mut backtrack;
+    let mut backtrack_subtree;
+
+    let mut value = Amount::ZERO;
+
+    let mut current_waste: SignedAmount = SignedAmount::ZERO;
+    let mut best_waste = SignedAmount::MAX_MONEY;
+
+    let mut index_selection: Vec<usize> = vec![];
+    let mut best_selection: Option<Vec<usize>> = None;
+
+    let upper_bound = target.checked_add(cost_of_change)?;
+
+    // Creates a tuple of (effective_value, waste, weighted_utxo)
+    let mut w_utxos: Vec<(Amount, SignedAmount, &WeightedUtxo)> = weighted_utxos
+        .iter()
+        // calculate effective_value and waste for each w_utxo.
+        .map(|wu| (wu.effective_value(fee_rate), wu.waste(fee_rate, long_term_fee_rate), wu))
+        // remove utxos that either had an error in the effective_value or waste calculation.
+        .filter(|(eff_val, waste, _)| eff_val.is_some() && waste.is_some())
+        // unwrap the option type since we know they are not None (see previous step).
+        .map(|(eff_val, waste, wu)| (eff_val.unwrap(), waste.unwrap(), wu))
+        // filter out all effective_values that are negative.
+        .filter(|(eff_val, _, _)| eff_val.is_positive())
+        // all utxo effective_values are now positive (see previous step) - cast to unsigned.
+        .map(|(eff_val, waste, wu)| (eff_val.to_unsigned().unwrap(), waste, wu))
```

Comment:
I don’t understand why you would change the type. What’s the advantage of casting to unsigned here?

Reply:
It may be possible to leave this as signed.  Since it's a positive value, I am not sure if there's a reason to cast this to Signed, although I see your point that maybe the computation can carry through with unsigned.  Will investigate further.